model/flow/header.go

```diff
@@ -72,54 +70,10 @@ func (h Header) Fingerprint() []byte {
 	return fingerprint.Fingerprint(h.Body())
 }
 
-var mutexHeader sync.Mutex
-var previdHeader Identifier
-var prevHeader Header
-
 // ID returns a unique ID to singularly identify the header and its block
 // within the flow system.
 func (h Header) ID() Identifier {
-	// NOTE: this is just a sanity check to make sure that we don't get
-	// different block IDs if someone forgets to use UTC timestamps
-	if h.Timestamp.Location() != time.UTC {
-		h.Timestamp = h.Timestamp.UTC()
-	}
-
-	mutexHeader.Lock()
-
-	// unlock at the return
-	defer mutexHeader.Unlock()
-
-	// compare these elements individually
-	if prevHeader.ParentVoterIndices != nil &&
-		prevHeader.ParentVoterSigData != nil &&
-		prevHeader.ProposerSigData != nil &&
-		len(h.ParentVoterIndices) == len(prevHeader.ParentVoterIndices) &&
-		len(h.ParentVoterSigData) == len(prevHeader.ParentVoterSigData) &&
-		len(h.ProposerSigData) == len(prevHeader.ProposerSigData) {
-
-		if h.ChainID == prevHeader.ChainID &&
-			h.Timestamp == prevHeader.Timestamp &&
-			h.Height == prevHeader.Height &&
-			h.ParentID == prevHeader.ParentID &&
-			h.View == prevHeader.View &&
-			h.PayloadHash == prevHeader.PayloadHash &&
-			bytes.Equal(h.ProposerSigData, prevHeader.ProposerSigData) &&
-			bytes.Equal(h.ParentVoterIndices, prevHeader.ParentVoterIndices) &&
-			bytes.Equal(h.ParentVoterSigData, prevHeader.ParentVoterSigData) &&
-			h.ProposerID == prevHeader.ProposerID {
-
-			// cache hit, return the previous identifier
-			return previdHeader
-		}
-	}
-
-	previdHeader = MakeID(h)
-
-	// store a reference to the Header entity data
-	prevHeader = h
-
-	return previdHeader
+	return MakeID(h)
```

Comment:
how many ID() and Checksum() calls do we make on each Header object?  We can just precompute and store the hash in the struct?

Reply:
computing the hash should have pretty much the same cost as comparing the fields, so the benefit unknown.  let me hack it with a sync.Once and run the benchmark to see if result is even measurable.

Reply:
hacked in header caching. (even without locks) it does not seem to affect performance on `ComputeBlock` tests:
```
name                             old us/tx      new us/tx      delta
ComputeBlock/1/cols/16/txes-8         593 ± 1%       588 ± 2%  -0.73%  (p=0.007 n=9+8)
ComputeBlock/1/cols/32/txes-8         561 ± 3%       561 ± 6%    ~     (p=0.347 n=10+9)
ComputeBlock/1/cols/64/txes-8         537 ± 2%       542 ± 3%    ~     (p=0.163 n=9+10)
ComputeBlock/1/cols/128/txes-8        542 ± 2%       544 ± 2%    ~     (p=0.847 n=9+9)
ComputeBlock/4/cols/16/txes-8         538 ± 1%       542 ± 1%  +0.78%  (p=0.045 n=10+10)
ComputeBlock/4/cols/32/txes-8         546 ± 2%       545 ± 3%    ~     (p=0.703 n=9+10)
ComputeBlock/4/cols/64/txes-8         546 ± 4%       552 ± 4%    ~     (p=0.197 n=9+9)
ComputeBlock/4/cols/128/txes-8        536 ± 3%       538 ± 7%    ~     (p=0.794 n=10+9)
ComputeBlock/16/cols/16/txes-8        541 ± 4%       533 ± 1%    ~     (p=0.059 n=10+10)
ComputeBlock/16/cols/32/txes-8        532 ± 2%       531 ± 2%    ~     (p=0.734 n=10+9)
ComputeBlock/16/cols/64/txes-8        535 ± 1%       532 ± 2%    ~     (p=0.249 n=9+10)
ComputeBlock/16/cols/128/txes-8       527 ± 2%       520 ± 1%  -1.39%  (p=0.007 n=9+10)
```